goaviatrix/fqdn.go

```diff
@@ -557,3 +562,74 @@ func (c *Client) DisableFQDNPassThrough(gw *Gateway) error {
 
 	return c.PostAPI(action, data, BasicCheck)
 }
+
+func (c *Client) AddFQDNTagRule(fqdn *FQDN) error {
+	action := "add_fqdn_policies_to_tag"
+
+	policies, err := json.Marshal(fqdn.DomainList)
+	if err != nil {
+		return fmt.Errorf("could not marshal fqdn domain: %v", err)
+	}
+
+	return c.PostAPI(action, struct {
+		Action   string `form:"action"`
+		CID      string `form:"CID"`
+		TagName  string `form:"tag_name"`
+		Policies string `form:"policies"`
+	}{
+		Action:   action,
+		CID:      c.CID,
+		TagName:  fqdn.FQDNTag,
+		Policies: string(policies),
+	}, BasicCheck)
+}
+
+func (c *Client) GetFQDNTagRule(fqdn *FQDN) (*FQDN, error) {
+	foundFQDN, err := c.ListDomains(&FQDN{FQDNTag: fqdn.FQDNTag})
+	if err != nil {
+		return nil, fmt.Errorf("could not list fqdn domains: %v", err)
+	}
+	domain := fqdn.DomainList[0]
+	found := false
+	for _, d := range foundFQDN.DomainList {
+		if d.FQDN == domain.FQDN && d.Protocol == domain.Protocol && d.Port == domain.Port && d.Verdict == domain.Verdict {
+			found = true
+		}
+	}
+	if !found {
+		return nil, ErrNotFound
+	}
+
+	return fqdn, nil
+}
+
+func (c *Client) DeleteFQDNTagRule(fqdn *FQDN) error {
+	action := "delete_fqdn_policies_to_tag"
+
+	policies, err := json.Marshal(fqdn.DomainList)
+	if err != nil {
+		return fmt.Errorf("could not marshal fqdn domain: %v", err)
+	}
+
+	return c.PostAPI(action, struct {
+		Action   string `form:"action"`
+		CID      string `form:"CID"`
+		TagName  string `form:"tag_name"`
+		Policies string `form:"policies"`
+	}{
+		Action:   action,
+		CID:      c.CID,
+		TagName:  fqdn.FQDNTag,
+		Policies: string(policies),
+	}, func(action string, reason string, ret bool) error {
+		if !ret {
+			// Tried to delete a rule that did not exist, we don't need to fail the apply.
+			if strings.Contains(reason, "the following rules were not found") {
+				return nil
```

Comment:
We should not probably return ErrNotFound here and ignore the error for destroy code.

The acceptance usually checks for it to confirm if it has been destroyed

Reply:
> ...should not probably return ErrNotFound here...

Do you mean we _should_ return ErrNotFound then just ignore it when calling the function?

Reply:
Yeah..that's what I meant. This current approach works as well.